sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala

```diff
@@ -0,0 +1,382 @@
+    }.asInstanceOf[T]
+  }
+
+  // Applies `outputMap` attribute mapping on attributes of `newExpressions` and merges them into
+  // `cachedExpressions`. Returns the merged expressions and the attribute mapping from the new to
+  // the merged version that can be propagated up during merging nodes.
+  private def mergeNamedExpressions(
+      newExpressions: Seq[NamedExpression],
+      outputMap: AttributeMap[Attribute],
+      cachedExpressions: Seq[NamedExpression]) = {
+    val mergedExpressions = ListBuffer[NamedExpression](cachedExpressions: _*)
+    val newOutputMap = AttributeMap(newExpressions.map { ne =>
+      val mapped = mapAttributes(ne, outputMap)
+      val withoutAlias = mapped match {
+        case Alias(child, _) => child
+        case e => e
+      }
+      ne.toAttribute -> mergedExpressions.find {
+        case Alias(child, _) => child semanticEquals withoutAlias
+        case e => e semanticEquals withoutAlias
+      }.getOrElse {
+        mergedExpressions += mapped
+        mapped
+      }.toAttribute
+    })
+    (mergedExpressions.toSeq, newOutputMap)
+  }
+
+  // Only allow aggregates of the same implementation because merging different implementations
+  // could cause performance regression.
+  private def supportedAggregateMerge(newPlan: Aggregate, cachedPlan: Aggregate) = {
+    val newPlanAggregateExpressions = newPlan.aggregateExpressions.flatMap(_.collect {
+      case a: AggregateExpression => a
+    })
+    val cachedPlanAggregateExpressions = cachedPlan.aggregateExpressions.flatMap(_.collect {
+      case a: AggregateExpression => a
+    })
+    val newPlanSupportsHashAggregate = Aggregate.supportsHashAggregate(
+      newPlanAggregateExpressions.flatMap(_.aggregateFunction.aggBufferAttributes))
+    val cachedPlanSupportsHashAggregate = Aggregate.supportsHashAggregate(
+      cachedPlanAggregateExpressions.flatMap(_.aggregateFunction.aggBufferAttributes))
+    newPlanSupportsHashAggregate && cachedPlanSupportsHashAggregate ||
+      newPlanSupportsHashAggregate == cachedPlanSupportsHashAggregate && {
+        val newPlanSupportsObjectHashAggregate =
+          Aggregate.supportsObjectHashAggregate(newPlanAggregateExpressions)
+        val cachedPlanSupportsObjectHashAggregate =
+          Aggregate.supportsObjectHashAggregate(cachedPlanAggregateExpressions)
+        newPlanSupportsObjectHashAggregate && cachedPlanSupportsObjectHashAggregate ||
+          newPlanSupportsObjectHashAggregate == cachedPlanSupportsObjectHashAggregate
+      }
+  }
+
+  // Second traversal replaces `ScalarSubqueryReference`s to either
+  // `GetStructField(ScalarSubquery(CTERelationRef to the merged plan)` if the plan is merged from
+  // multiple subqueries or `ScalarSubquery(original plan)` if it isn't.
+  private def removeReferences(
+      plan: LogicalPlan,
+      cache: ListBuffer[Header]): (LogicalPlan, Seq[CTERelationDef]) = {
+    val subqueryCTEs = mutable.Map.empty[Int, CTERelationDef]
```

Comment:
Do we really need to use CTE? It seems like we don't leverage any CTE feature. We just wait for the optimizer to inline all the CTE relations and expect the planner to trigger subquery reuse.

Reply:
That is basically true. This is the relevant conversation we had with @sigmod: https://github.com/apache/spark/pull/32298#discussion_r627800694

Briefly, this PR didn't depend on CTEs originally, but then `MergeScalarSubqueries` evolved to a general rule that might be followed by column pruning. As column pruning might alter the subqueries, subquery reuse might not kick in. So we needed a way to keep merged subqueries untouched till the end of optimization. First I used a custom `CommonScalarSubqueries` root node with `ScalarSubqueryReference` expressions, but then the implementation was changed to use general CTE nodes.

Reply:
column pruning does not cross the subquery boundaries, I'm not sure what to worry about here. Plus, CTE can do column pruning and filter pushdown now.

Reply:
column pruning does not cross the subquery boundaries, I'm not sure what to worry about here. Plus, CTE can do column pruning and filter pushdown now.

Reply:
I'm fine with both approaches. cc @sigmod 

I guess CTE column pruning and filter pushdown takes all CTE references into into account so it should not cause any issue...

Reply:
Ah I see your point. Using CTE guarantees that we can trigger subquery reuse, but my point is it should be guaranteed anyway. The same query plan in different scalar subqueries will remain the same AFAIK. @sigmod are you aware of any potential issues?

Reply:
after a second thought, we may do column pruning for `GetStructField(SubqueryExpression(plan))` in the future, using CTE is more future-proof